go/vt/tabletmanager/rpc_replication.go

```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
 	return false
 }
 
-func (agent *ActionAgent) enableSemiSync(master bool) error {
+func (agent *ActionAgent) setSemiSync(tabletType topodatapb.TabletType) error {
+	if !*enableSemiSync {
+		// Semi-sync handling is not enabled.
+		return nil
+	}
+
 	// Only enable if we're eligible for becoming master (REPLICA type).
 	// Ineligible slaves (RDONLY) shouldn't ACK because we'll never promote them.
-	if !isMasterEligible(agent.Tablet().Type) {
-		return nil
+	if !isMasterEligible(tabletType) {
+		return agent.MysqlDaemon.SetSemiSyncEnabled(false, false)
 	}
 
 	// Always enable slave-side since it doesn't hurt to keep it on for a master.
 	// The master-side needs to be off for a slave, or else it will get stuck.
-	return agent.MysqlDaemon.SetSemiSyncEnabled(master, true)
+	return agent.MysqlDaemon.SetSemiSyncEnabled(tabletType == topodatapb.TabletType_MASTER, true)
+}
+
+func (agent *ActionAgent) setSemiSyncAndFixReplication(tabletType topodatapb.TabletType) error {
```

Comment:
If you name the method above "fixSemiSync" then this could be very conveniently named "fixSemiSyncAndReplication". It would be much better.